core/src/main/java/org/jdbi/v3/core/mapper/reflect/ImmutableMapper.java

```diff
@@ -0,0 +1,223 @@
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Type;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+import static org.jdbi.v3.core.mapper.reflect.ReflectionMapperUtil.findColumnIndex;
+import static org.jdbi.v3.core.mapper.reflect.ReflectionMapperUtil.getColumnNames;
+
+/**
+ * Maps to an Interface with googles Immutable libary.
+ */
+public class ImmutableMapper<T> implements RowMapper<T> {
+
+    static final String DEFAULT_PREFIX = "";
+
+    /**
+     * Returns a mapper factory that maps to the given immutable class
+     *
+     * @param type the mapped class
+     * @return a mapper factory that maps to the given immutable class
+     */
+    public static RowMapperFactory factory(Class<?> type) {
+        return RowMapperFactory.of(type, ImmutableMapper.of(type));
+    }
+
+    /**
+     * Returns a mapper factory that maps to the given immutable class
+     *
+     * @param type the mapped class
+     * @param prefix the column name prefix for each mapped immutable property
+     * @return a mapper factory that maps to the given immutable class
+     */
+    public static RowMapperFactory factory(Class<?> type, String prefix) {
+        return RowMapperFactory.of(type, ImmutableMapper.of(type, prefix));
+    }
+
+    /**
+     * Returns a mapper for the given immutable class
+     *
+     * @param <T> the type to find the mapper for
+     * @param type the mapped class
+     * @return a mapper for the given immutable class
+     */
+    public static <T> RowMapper<T> of(Class<T> type) {
+        return ImmutableMapper.of(type, DEFAULT_PREFIX);
+    }
+
+    /**
+     * Returns a mapper for the given immutable class
+     *
+     * @param <T> the type to find the mapper for
+     * @param type the mapped class
+     * @param prefix the column name prefix for each mapped immutable property
+     * @return a mapper for the given immutable class
+     */
+    public static <T> RowMapper<T> of(Class<T> type, String prefix) {
+        return new ImmutableMapper<>(type, prefix);
+    }
+
+    private final Class<? extends T> type;
+    private final Class<? extends T> implementation;
+    private final Object builder;
+    private final String prefix;
+    private final ImmutableInfo info;
+    private final Map<PropertyDescriptor, ImmutableMapper<?>> nestedMappers = new ConcurrentHashMap<>();
+
+    private ImmutableMapper(Class<T> type, String prefix) {
+
+        this.type = type;
+        this.prefix = prefix;
+        this.implementation = (Class<? extends T>) type.getAnnotation(ImmutableImplementation.class).value();
+
+        try {
+            builder = this.implementation.getMethod("builder").invoke(null);
+        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
```

Comment:
Could this use the new `MethodHandles` approach to avoid exception wrapping and other ickiness of the old reflection API?

Reply:
Not here because I can not know the returntype